rust/ql/lib/codeql/rust/controlflow/internal/ControlFlowGraphImpl.qll

```diff
@@ -532,3 +569,164 @@ class YieldExprTree extends StandardPostOrderTree instanceof YieldExpr {
 class YeetExprTree extends StandardPostOrderTree instanceof YeetExpr {
   override AstNode getChildNode(int i) { i = 0 and result = super.getExpr() }
 }
+
+/**
+ * Provides `ControlFlowTree`s for patterns.
+ *
+ * Since patterns destruct values, they are modeled in pre-order, except for
+ * `OrPat`s and `IdentPat`s.
```

Comment:
Why are ident and or-patterns different?

Reply:
Because in `e @ pat` we only "evalute" `e` when `pat` matches. And in `pat1 | pat2` we need it to be post-order in order to get the correct dominance information needed for guards.